modules/core/src/main/java/org/apache/ignite/internal/util/distributed/DistributedProcess.java

```diff
@@ -0,0 +1,388 @@
+            if (msg0 instanceof SingleNodeMessage && ((SingleNodeMessage)msg0).type() == type.ordinal()) {
+                SingleNodeMessage<R> msg = (SingleNodeMessage<R>)msg0;
+
+                if (msg.type() == type.ordinal())
+                    onSingleNodeMessageReceived(msg, nodeId);
+            }
+        });
+
+        ctx.event().addDiscoveryEventListener((evt, discoCache) -> {
+            UUID leftNodeId = evt.eventNode().id();
+
+            for (Process p : processes.values()) {
+                p.initFut.listen(fut -> {
+                    if (F.eq(leftNodeId, p.crdId)) {
+                        ClusterNode crd = coordinator();
+
+                        if (crd == null) {
+                            onAllServersLeft();
+
+                            return;
+                        }
+
+                        p.crdId = crd.id();
+
+                        if (crd.isLocal())
+                            initCoordinator(p, discoCache.version());
+
+                        if (!ctx.clientNode())
+                            p.resFut.listen(f -> sendSingleMessage(p, crd));
+                    }
+                    else if (F.eq(ctx.localNodeId(), p.crdId)) {
+                        boolean rmvd, isEmpty;
+
+                        synchronized (mux) {
+                            rmvd = p.remaining.remove(leftNodeId);
+
+                            isEmpty = p.remaining.isEmpty();
+                        }
+
+                        if (rmvd) {
+                            assert !p.singleMsgs.containsKey(leftNodeId);
+
+                            if (isEmpty)
+                                finishProcess(p);
+                        }
+                    }
+                });
+            }
+        }, EVT_NODE_FAILED, EVT_NODE_LEFT);
+    }
+
+    /**
+     * Starts distributed process.
+     *
+     * @param req Initial request.
+     * @param id Process id.
+     */
+    public void start(UUID id, I req) {
+        try {
+            InitMessage msg = new InitMessage<>(id, type.ordinal(), req);
+
+            ctx.discovery().sendCustomEvent(msg);
+        }
+        catch (IgniteCheckedException e) {
+            log.warning("Unable to start process.", e);
+        }
+    }
+
+    /**
+     * Initiates process coordinator.
+     *
+     * @param p Process.
+     * @param topVer Topology version.
+     */
+    private void initCoordinator(Process p, AffinityTopologyVersion topVer) {
+        synchronized (mux) {
+            if (p.initCrdFut.isDone())
+                return;
+
+            assert p.remaining.isEmpty();
+
+            p.remaining.addAll(F.viewReadOnly(ctx.discovery().serverNodes(topVer), F.node2id()));
+
+            p.initCrdFut.onDone();
+        }
+    }
+
+    /**
+     * Sends single node message to coordinator.
+     *
+     * @param p Process.
+     * @param crd Coordinator node to send message.
+     */
+    private void sendSingleMessage(Process p, ClusterNode crd) {
```

Comment:
Do we really need `crd` here since the `Process` already has its `crdId`?

Reply:
Yes, the parameter is redundant. I'll refactor it soon.